pkg/api/steve/collections/rbac.go

```diff
@@ -0,0 +1,315 @@
+			err = h.createClusterRole(collectionRoleBinding.RoleRef.Name, collectionRules, clusterRole)
+			if err != nil {
+				return nil, err
+			}
+			return rb, nil
+		}
+		if _, ok := collectionClusterRole.GetAnnotations()[clusterRoleOwnerAnnotation]; ok { // owned by roletemplate controllers, don't update it
+			return rb, nil
+		}
+		err = h.compareAndUpdateClusterRole(collectionClusterRole, collectionRules, clusterRole)
+		if err != nil {
+			return nil, err
+		}
+		return rb, nil
+	}
+	collectionRole, err := h.rolesCache.Get(namespace.Name, collectionRoleBinding.RoleRef.Name)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return nil, err
+	}
+	if collectionRole == nil {
+		err = h.createRole(collectionRoleBinding.RoleRef.Name, namespace.Name, collectionRules, role)
+		if err != nil {
+			return nil, err
+		}
+		return rb, nil
+	}
+	err = h.compareAndUpdateRole(collectionRole, collectionRules, role)
+	if err != nil {
+		return nil, err
+	}
+
+	return rb, nil
+}
+
+func GetCollectionRules(apis APIResourceWatcher, rules []rbacv1.PolicyRule) ([]rbacv1.PolicyRule, error) {
+	collectionRules := []rbacv1.PolicyRule{}
+	validRules := []rbacv1.PolicyRule{}
+	for _, rule := range rules {
+		if len(rule.ResourceNames) != 0 {
+			continue
+		}
+		if len(rule.NonResourceURLs) != 0 {
+			continue
+		}
+		for _, verb := range rule.Verbs {
+			if verb == "*" || verb == "list" {
+				validRules = append(validRules, rule)
+			}
+		}
+	}
+	if len(validRules) == 0 {
+		return collectionRules, nil
+	}
+	if len(apis.List()) == 0 {
+		return collectionRules, ErrNotReady
+	}
+	for _, rule := range validRules {
+		for _, apiGroup := range rule.APIGroups {
+			for _, resource := range rule.Resources {
+				_, ok := apis.Get(resource, apiGroup)
+				if ok {
+					if len(collectionRules) == 0 {
+						collectionRules = []rbacv1.PolicyRule{
+							{
+								APIGroups: []string{Group},
+								Verbs:     []string{"list"},
+							},
+						}
+					}
+					if apiGroup != "" {
+						resource = apiGroup + "." + resource
+					}
+					collectionRules[0].Resources = append(collectionRules[0].Resources, resource)
+					continue
+				}
+				if resource == "*" {
+					if apiGroup == "*" {
+						collectionRules = []rbacv1.PolicyRule{
+							{
+								APIGroups: []string{Group},
+								Verbs:     []string{"list"},
+								Resources: []string{"*"},
+							},
+						}
+						return collectionRules, nil
+					}
+					apis := apis.List()
+					for _, api := range apis {
+						if api.Group == apiGroup {
+							if len(collectionRules) == 0 {
+								collectionRules = []rbacv1.PolicyRule{
+									{
+										APIGroups: []string{Group},
+										Verbs:     []string{"list"},
+									},
+								}
+							}
+							collectionRules[0].Resources = append(collectionRules[0].Resources, api.Name)
+						}
+					}
+					if len(collectionRules) > 0 {
+						sort.Strings(collectionRules[0].Resources)
+					}
+					return collectionRules, nil
```

Comment:
I might be missing something, but this return feels premature to me. We could have a case where there are `*` resources on two api groups. This seems like it will only add the resources for one of those and return early.

Reply:
I think you're right, will fix

Reply:
Resolving, approach changed, file no longer exists.